server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java

```diff
@@ -571,4 +583,98 @@ public MetadataUpdateCount getUpdateCount() {
     return metadataUpdateCount.get();
   }
 
+  /**
+   * This is called when metadata file list is determined to be out of sync withe in-memory file
+   * list. Firstly the operation log is dumped. Then metadata is compared again with memory. If
+   * still out of sync, then recovery is performed dependent on the action: 1) log: nothing else is
+   * done 2) logsync: memory is synched with the metadata IFF the operation log agrees with one or
+   * the other 3) metasync: metadata files are reloaded into memory IFF the operation log agrees
+   * with metadata 4) memsync: metadata is reset with the memory files IFF the operation log agrees
+   * with memory
+   *
+   * @param context The server context
+   */
+  public void handleMetadataDiff(ServerContext context) {
+    LogSync action =
+        LogSync.valueOf(tablet.getTableConfiguration().get(Property.TABLE_OPERATION_LOG_RECOVERY));
+    synchronized (tablet) {
+      // always log the operation log regardless of the requested action
+      log.error("Operation log: " + tabletLog.dumpLog());
+
+      // rescan for the tablet metadata
+      var tabletMeta =
+          context.getAmple().readTablet(tablet.getExtent(), TabletMetadata.ColumnType.FILES);
+      if (tabletMeta == null) {
+        String msg = "Tablet " + tablet.getExtent() + " not found in metadata";
+        log.error(msg);
+      } else {
+        Set<StoredTabletFile> metadata = new HashSet<>(tabletMeta.getFiles());
+        Set<StoredTabletFile> memory = new HashSet<>(datafileSizes.keySet());
+        Set<StoredTabletFile> expected = tabletLog.getExpectedFiles();
+
+        // verify we are still out of sync
+        if (metadata.equals(memory)) {
+          log.debug("Metadata and in-memory file list are back in-sync: " + metadata);
+          if (!expected.equals(memory)) {
+            log.error(
+                "Resetting operation log " + expected + " with metadata and memory " + memory);
+            tabletLog.reset(memory);
+          }
+        } else if (!action.equals(LogSync.log)) {
+          if (action.equals(LogSync.logsync)) {
+            if (expected.equals(memory)) {
+              action = LogSync.memsync;
+            } else if (expected.equals(metadata)) {
+              action = LogSync.metasync;
+            } else {
+              log.error("Not synching files because the operation log " + expected
+                  + " does not agree with metadata " + metadata + " nor memory " + memory);
```

Comment:
```suggestion
              log.error("Not syncing files because the operation log {}"
                  + " does not agree with metadata {}  nor memory {}", expected, metadata, memory);
```